### PR TITLE
Max-allowed Pods for nodes

### PR DESCRIPTION
**Pull request by islinwb**
for #88
Firmament part: https://github.com/Huawei-PaaS/firmament/pull/3

### CODE REVIEW

**Comment by m1093782566** [comment]
Great @islinwb 

It looks promising though it's still WIP.

**Comment by islinwb** [comment]
@m1093782566 @shivramsrivastava The PR's ready for reviewing.

**Comment by m1093782566** [comment]
@islinwb 

Should we merge the firmament changes first?

**Comment by islinwb** [comment]
@m1093782566  Yes. The added e2e test case can pass only after merging https://github.com/Huawei-PaaS/firmament/pull/3.

File: pkg/k8sclient/nodewatcher.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ func (nw *NodeWatcher) parseNode(node *v1.Node, phase NodePhase) *Node {
 	memCap, _ := memCapQuantity.AsInt64()
 	memAllocQuantity := node.Status.Allocatable[v1.ResourceMemory]
 	memAlloc, _ := memAllocQuantity.AsInt64()
+	podCapQuantity := node.Status.Capacity[v1.ResourcePods]
```
Comment (m1093782566) [inline pkg/k8sclient/nodewatcher.go]: It's never really used by firmament?

File: pkg/k8sclient/nodewatcher.go [hunk 2]
```diff
@@ -115,6 +118,8 @@ func (nw *NodeWatcher) parseNode(node *v1.Node, phase NodePhase) *Node {
 		CPUAllocatable:   cpuAllocQuantity.MilliValue(),
 		MemCapacityKb:    memCap / bytesToKb,
 		MemAllocatableKb: memAlloc / bytesToKb,
+		PodCapacity:      podCapQuantity.Value(),
```
Comment (m1093782566) [inline pkg/k8sclient/nodewatcher.go]: It's never really used by firmament? We just pass the `PodAllocatable` to firmament?

**Comment by shivramsrivastava** [comment]
@islinwb Thanks for the PR.
I will review it in sometime.

For the E2E i would suggest to create an image and upload it [here](https://hub.docker.com/r/huaweifirmament/firmament/tags/), with a tag like ```'e2e'``` or ```'v0.3_beta'```. But the firmament side code needs to be reviewed and merged first.

Also shouldn't the test case modify capacity for nodes and check, i don't know if an update to node will set the capacity, or one needs to restart the ```kubelet``` of the node to change the capacity.

Because what i understands from the k8s documentation, the max pods limit is set at the time of bringing up the node, and later changes we need to restart the ```kubelet``` ???

If the later case is true, where we have to restart the nodes agent to make the capacity change.
Then we can simply the code on firmament side?
We don't have to consult the ```knowledgeable``` for the node capacity calculation ?

File: pkg/k8sclient/nodewatcher.go (hunk 2 above)
Comment (islinwb) [reply pkg/k8sclient/nodewatcher.go]: @m1093782566  Only use `PodAllocatable` now. I'll check whether `PodCapacity` should be used.

**Comment by islinwb** [comment]
@shivramsrivastava
> Because what i understands from the k8s documentation, the max pods limit is set at the time of bringing up the node, and later changes we need to restart the `kubelet` ???

Yes, kubelet should be restarted if we want to change `max-pods` of kubelet.

The e2e I added only tests whether the Firmament flag `max_multi_arcs_for_cpu` works (assuming `max-pods` greater than `max_multi_arcs_for_cpu` since their default value is 110 and 50 respectively).

When  `max-pods` < `max_multi_arcs_for_cpu`, for example, 40 and 50 respectively. The Firmament scheduler will create 40 ec_to_machine arc. However the Firmament scheduler fails to take pods like `kube-dns` scheduled by default scheduler into consideration. Problem occur when you create a deployment/rs with 40+ replicas.

**Comment by islinwb** [comment]
@shivramsrivastava @m1093782566  I've tested in my local VM with the modified Firmament code. The kubelet flags max-pods works.
There's a issue since the Firmament can't get the real-time pods number on nodes when other scheduler works at the same time: Firmament will try to create new pods (the pods number is increaing) when the intended replicas is greater than the allowed (there are already other pods not scheduled by Firmament) .

**Comment by m1093782566** [comment]
We don't need to consider other scheduler for the time being.

**Pull request closed without merging**